### PR TITLE
Change CLI commands and options (part 1)

### LINKED ISSUE #16033: review quarkus cli command / flag structure
current cli outline looks like this:

```
Usage: quarkus [-ehV] [--verbose] [COMMAND]
  -e, --errors    Produce execution error messages.
  -h, --help      Show this help message and exit.
  -V, --version   Print version information and exit.
      --verbose   Verbose mode.
Commands:
  build             Build your quarkus project
  clean             Clean current project
  create            Create a new quarkus project.
  create-jbang      Create a new quarkus jbang project.
  list              List installed (default) or installable extensions.
  platforms         List imported (default) or all available Quarkus platforms.
  add               Add extension(s) to current project.
  remove, rm        Remove an extension from this project.
  dev               Execute project in live coding dev mode
  create-extension  Creates the base of a Quarkus extension in different layout depending of the options and environment.
```

they currently can be grouped in following categories:

## project creation 

create, create-jbang, create-extension

imo. `create-jbang` should be a `--jbang` flag instead of its own command.

## platform/extension query/manipulation

list,add,remove/rm
platforms

kinda core to what we would like the cli to do but feels a bit weird we have list/add/remove/rm for extensions but then a separate platform command.

## dev time commands

build, clean, dev

force: they abstract away wether you are doing maven or gradle (in future jbang could be included here)

weakness: you cant express clean build - they always end up as two commands ..
missing remote dev

missing commands/ideas:

update - would check if quarkus has a newer release and update the corresponding build artifact. 

validate/lint - be able to see if using out-of-platform-bounds artifacts.

### LINKED ISSUE #16316: CLI: make it possible to specify additional quarkus config properties
ATM it's not possible to pass quarkus config properties, e.g. something like `qs build -p quarkus.package.type=uber-jar`.

See also https://picocli.info/#_maps

### LINKED ISSUE #16318: CLI: consider autocomplete support via TAB
I believe that most of the CLI users are used to use this feature.

See https://picocli.info/autocomplete.html.

### PR DESCRIPTION
**Pull request by ebullient**
Resolves #16033 
Resolves #16316
Resolves #16318
Resolves #16407

### CODE REVIEW

File: devtools/cli/src/main/java/io/quarkus/cli/Create.java
Comment (gsmet) [inline devtools/cli/src/main/java/io/quarkus/cli/Create.java:18]: Last time I was involved in the discussion, I said `create` was not explicit enough as I have no idea what it creates. I think this still stands.

Especially since we will have `create-extension`.

I thought we agreed on this? Did you rediscuss this again?

File: devtools/cli/src/main/java/io/quarkus/cli/QuarkusCli.java
Comment (gsmet) [inline devtools/cli/src/main/java/io/quarkus/cli/QuarkusCli.java:22]: Same here about `Add` and `Remove`, it's definitely not clear enough. You need to know what you are adding or removing. Especially since you have `CreateExtension`.

File: devtools/cli/src/main/java/io/quarkus/cli/QuarkusCli.java
Comment (gsmet) [inline devtools/cli/src/main/java/io/quarkus/cli/QuarkusCli.java:19]: So in the end, we decided against `qs`?

**Review by gsmet — DISMISSED** [review]
-/-

File: devtools/cli/src/main/java/io/quarkus/cli/QuarkusCli.java
Comment (gsmet) [reply devtools/cli/src/main/java/io/quarkus/cli/QuarkusCli.java:22]: Apart from clarity, what's also bothering me is that we paint ourselves into a corner in case we want to `add` other things later. And that doesn't seem that far fetched. Same for `list`. It's very generic and I'm pretty sure we will want to list other things at some point.

I would rather have `ListExtensions`, `AddExtension` and `RemoveExtension`.

File: devtools/cli/src/main/java/io/quarkus/cli/QuarkusCli.java
Comment (gsmet) [reply devtools/cli/src/main/java/io/quarkus/cli/QuarkusCli.java:19]: I'm fine with it btw, just wanted to check.

Comment (ebullient) [reply devtools/cli/src/main/java/io/quarkus/cli/QuarkusCli.java:19]: In the email thread (I think), it was stated we should just go with quarkus, as aliases are easy enough to add. 

Nothing has been "decided", this is literally an experiment so people can play with it (e.g. tab complete and help).

**Comment by Ladicek** [comment]
I don't have the full context here, but I'd like to point out one thing:

> If we have proper completion, I think it's better to have clarity rather than avoiding typing characters.

Note that a first-time user will _not_ have completion installed. Most likely even a second-time user. Considering completion is somewhat clunky to enable, I firmly believe that it's best to assume that only regular users have completion enabled and functioning. Irregular users, especially first-time users, do not. Completion is not a panacea. Commands must still be short and easy to type.

File: devtools/cli/src/main/java/io/quarkus/cli/Create.java
Comment (ebullient) [reply devtools/cli/src/main/java/io/quarkus/cli/Create.java:18]: `create-app` is easy enough to amend. I can do that today and we can see what it feels like.

File: devtools/cli/src/main/java/io/quarkus/cli/QuarkusCli.java
Comment (ebullient) [reply devtools/cli/src/main/java/io/quarkus/cli/QuarkusCli.java:22]: The internal class names are for reference only (and I tried to group them in the list of subcommands also, to show they are related).

**Comment by mkouba** [comment]
I agree with @Ladicek autocompletion is a nice-to-have but nothing we can rely on when designing the commands...

File: devtools/cli/src/main/java/io/quarkus/cli/Create.java
Comment (mkouba) [reply devtools/cli/src/main/java/io/quarkus/cli/Create.java:18]: +1 for create-app

File: devtools/cli/src/main/java/io/quarkus/cli/QuarkusCli.java
Comment (mkouba) [reply devtools/cli/src/main/java/io/quarkus/cli/QuarkusCli.java:22]: I like the idea of clear and consistent names, i.e. `qs list-extensions` and `ListExtensions`

**Comment by gsmet** [comment]
> Commands must still be short and easy to type.

How having unclear commands will help? You have to keep in mind that most of our users:
- don't create new projects daily
- don't add extensions to their project daily
- have other projects using other technologies
- are not in Quarkus mode always

So when you come back to your Quarkus app, it's far easier to understand that `create-app` creates an app and `add-extension` adds an extension rather than `create` and `add` which are not descriptive enough and you have no idea of what they do.

And again, what will you do when you will want to add/list/remove other things than extensions?

And how do you explain that `create` creates an app, that `create-extension` creates an extension but `add` adds an extension?

I don't want to make the CLI cumbersome to use but we need consistency and clarity. And I don't think having consistent names will be that much of a problem, it's not as if we would have veeeeerrryyyyy long commands.

**Comment by Ladicek** [comment]
I think 99.9% of users will want to use this CLI to create/work with Quarkus _applications_, so it should be optimized for this case. If there are people who use it to work with projects that are in fact Quarkus _extensions_, I think it's fine to hide that under a special subcommand. I'd say that `quarkus create` and `quarkus add` and `quarkus whatever` just work with Quarkus applications; `quarkus extension create` or `quarkus extension i-dont-know-what` are for Quarkus extensions. I see no point in degrading the "application" experience just to achieve consistency with the "extension" experience, which is much less often used. (Especially if you can achieve similar consistency by putting extension-related commands under `quarkus extension`.)

But here's another thing, vastly more important than what I wrote in the first paragraph. Designing user interface by committee never works. We can argue all time long, but in the end, I believe we should pick one person (preferably one with some user interface design experience :-) ) and let them have the final say. That's the only way (I'm aware of) to make user experience design actually work.

**Comment by ebullient** [comment]
Agree, but I like input, too.
And I've been down the CLI design rathole several times.

**Comment by maxandersen** [comment]
> Designing user interface by committee never works. We can argue all time long, but in the end, I believe we should pick one person (preferably one with some user interface design experience :-)

agreed. cli was not to be designed by comittee - the cli work been dormant waiting on someone to have the bandwidth to do it - we got lots of the big items outlined and now @ebullient stepped up working on it. sharing the cli "ui" via PR's is imo good to get feedback/catch gotchas. @ebullient can ignore it all or what I think she will do - reflect and pick the best choice(s)!

**Comment by maxandersen** [comment]
@ebullient how much in this PR is supposed to work ? 

```
qssnap create
Unmatched argument at index 0: 'create'
Did you mean: create-cli or create-extension or remove-extension?
❯ qssnap create create-app
Unmatched argument at index 0: 'create'
Did you mean: create-cli or create-extension or remove-extension?
```

but `qssnap` does show:

```
qssnap
Usage: quarkus [-ehV] [--verbose] [COMMAND]
  -e, --errors    Display error messages.
  -h, --help      Show this help message and exit.
  -V, --version   Print version information and exit.
      --verbose   Verbose mode.
Commands:
  create-app            Create a Quarkus application project
  create-cli            Create a Quarkus command-line project
  create-extension      Create a Quarkus extension project
  build                 Build the current project
  dev                   Run the current project in dev mode (live coding)
  list-extension        List platform(s) and extensions
  add-extension         Add extension(s) to current project.
  remove-extension, rm  Remove an extension from this project.
  generate-completion   Generate bash/zsh completion script for quarkus.
```

so supposedly `create-app` is there...but isn't found...

**Comment by maxandersen** [comment]
funny enough `qssnap create-app --help` actually shows help :)

**Comment by maxandersen** [comment]
Just leaving comment based on chat conversation with @ebullient:

*-extension (list, remove, etc.) seems "too different". suggestion is to use extension-* or simply extension * and like there is a `rm` shortcut we can have that for the others to make app dev workflow as simple as possible.

**Comment by ebullient** [comment]
> funny enough `qssnap create-app --help` actually shows help :)

yep. this is all about help. should dump args if you try invoking.

**Comment by maxandersen** [comment]
It prints something that doesn't have a tostring :)

**Comment by ia3andy** [comment]
Hey hey,
I've been playing a bit with the new layout and here is my feedback.

Here is the current help:
```
Usage: quarkus [-ehV] [--verbose] [COMMAND]
  -e, --errors    Display error messages.
  -h, --help      Show this help message and exit.
  -V, --version   Print version information and exit.
      --verbose   Verbose mode.
Commands:
  create-app            Create a Quarkus application project
  create-cli            Create a Quarkus command-line project
  create-extension      Create a Quarkus extension project
  build                 Build the current project
  dev                   Run the current project in dev mode (live coding)
  list-extension        List platform(s) and extensions
  add-extension         Add extension(s) to current project.
  remove-extension, rm  Remove an extension from this project.
  generate-completion   Generate bash/zsh completion script for quarkus.
```

- Overall it feels really nice to have it clear and tidy!

- While `list-extension`, `add-extension` and `remove-extension` are clearer, I preferred shorter add/remove/rm/list for usage, maybe we could keep the aliases? 
```shell
q add my-ext
q rm my-ext
q remove my-ext
q list
```

SIDE NOTE: Wouldn't `search` be more revelent than `list`?


- `create-extension` feels wrong next to `create-cli` and `create-app`. An extension is not a type of Quarkus app. Also if we have `create-app` to "Create a Quarkus application project" but a `cli` is also a Quarkus Application project. 
	- Maybe we should have `create-rest`, `create-reactive-rest` or such so we don't feel that `cli` is different?
	- Else just keeping `create-app` and `create-extension` then have some sub-commands/shortcuts like `create-app cli` or `create-app rest` would be more consistant IMO
	- At least we should put the `create-extension` under the `remove-extension` to show it's not part of the same group.

Thanks for this and I know it's a pain to make everyone happy, in the end, you are making it ;)

**Comment by ebullient** [comment]
>     * While `list-extension`, `add-extension` and `remove-extension` are clearer, I preferred shorter add/remove/rm/list for usage, maybe we could keep the aliases?
> 
> 
> ```shell
> q add my-ext
> q rm my-ext
> q remove my-ext
> q list
> ```

This + @maxandersen's comment gives me an idea... 

> SIDE NOTE: Wouldn't `search` be more revelent than `list`?

possibly...

>     * `create-extension` feels wrong next to `create-cli` and `create-app`. An extension is not a type of Quarkus app. Also if we have `create-app` to "Create a Quarkus application project" but a `cli` is also a Quarkus Application project.

Here is my reasoning: 
1) `create-extension`: we want to encourage people to make quarkiverse extensons, and the layout for those is persnickety.
2) `create-cli`: Hey, did you know you can use Quarkus to make command line apps, and not just long-running server things? (insert jbang as default build tool here -- this is a different animal, and is a use case that is "new")
3) `create-app`: everything else.

I like the input. ;)

**Comment by ia3andy** [comment]
> [cli-2021-04-30.txt (1)](https://github.com/quarkusio/quarkus/files/6407976/cli-2021-04-30.txt)
> with better dev/build: [cli-2021-04-30.txt (2)](https://github.com/quarkusio/quarkus/files/6408111/cli-2021-04-30.txt)
> 
> ```
> $ quarkus --help
> Usage: quarkus [-BehV] [--verbose] [COMMAND]
>   -B, --batch-mode   run command in batch mode
>   -e, --errors       Display error messages.
>   -h, --help         Show this help message and exit.
>   -V, --version      Print version information and exit.
>       --verbose      Verbose mode.
> Commands:
>   create-app            Create a Quarkus application project
>   create-cli            Create a Quarkus command-line project
>   create-extension      Create a Quarkus extension project
>   build                 Build the current project
>   dev                   Run the current project in dev mode (live coding)
>   extensions, ext       List, add, and remove extensions
>     add                 Add extension(s) to current project
>     list, ls            List platform(s) and extensions
>     remove, rm          Remove an extension from this project
>   generate-completion   Generate bash/zsh completion script for quarkus.
> ```
> 
> @ia3andy .. `quarkus ext add ... `
> Did you see my other comments re: why those three create options? WDYT?

Great!

Few suggestions:
- We could move the `create-extension` to `extension create`? or at least move it just under the `extensions`
- IMO `extension` singular would be better

**Comment by maxandersen** [comment]
> * We could move the `create-extension` to `extension create`? or at least move it just under the `extensions`

how about doing like this in name of consistency:

```
Commands:
  create                Create Quarkus projects
    app                 Create a Quarkus application project
    cli                 Create a Quarkus command-line project
    extension           Create a Quarkus extension project
  build                 Build the current project
  dev                   Run the current project in dev mode (live coding)
  extensions, ext       List, add, and remove extensions
    add                 Add extension(s) to current project
    list, ls            List platform(s) and extensions
    remove, rm          Remove an extension from this project
  completion   Generate bash/zsh completion script for quarkus.
```

- have sub commands under "create" instead of 3 highlevel ones ? Is/was there a reason to have it as top levels?

- use "completion" as name as that seem to what others use, i.e. kubectl and jbang :)

**Comment by ebullient** [comment]
> how about doing like this in name of consistency:
> 
> ```
> Commands:
>   create                Create Quarkus projects
>     app                 Create a Quarkus application project
>     cli                 Create a Quarkus command-line project
>     extension           Create a Quarkus extension project
>```
> - have sub commands under "create" instead of 3 highlevel ones ?

I coded that yesterday! will post update today (with more for build/dev)..

> ```
>   completion   Generate bash/zsh completion script for quarkus.
> ```
> - use "completion" as name as that seem to what others use, i.e. kubectl and jbang :)

I will look... generate-completion is used with kubectl and others (helm?), and is a picocli built-in.

**Comment by ebullient** [comment]
I haven't renamed generate-completion yet, but properties/parameters added to `dev` and `build` 
These still don't do anything but collect the values you type! And print out help. ;)
So.. provided you grok the limitations, bang away and test command line completion, etc.

[cli-2021-05-05.txt](https://github.com/quarkusio/quarkus/files/6431557/cli-2021-05-05.txt)
@maxandersen

File: devtools/cli/src/main/java/io/quarkus/cli/Build.java
Comment (mkouba) [inline devtools/cli/src/main/java/io/quarkus/cli/Build.java:17]: I would almost say that this should be `true` by default. I know that it's a special goal/task in maven/gradle but from my exp it's always good to start the build with `mvn clean ...`. Or at least it would help if it would be possible to say _"always clean my project during build so that I can do `quarkus build`"_ once and the option would the persisted somewhere...

Comment (ebullient) [reply devtools/cli/src/main/java/io/quarkus/cli/Build.java:17]: interesting thought. I think I could make this a toggle-able thing... (the --clean --no-clean variety? )

**Comment by ebullient** [comment]
Ok. I've updated lots of flags to be toggles, which really changed the look/feel of `dev` and `build` .. and I think improves clarity quite a bit. See what you think: 

[cli-2021-05-06.txt](https://github.com/quarkusio/quarkus/files/6435680/cli-2021-05-06.txt)

Again, this can be built and "run" .. such that you can play with code completion and the impact of options to see how it feels.

**Comment by maxandersen** [comment]
Adding the properties feature is nice for the usecase of books/tutorials as you can then just share a command line and have a generated projected with everything but the code setup to go. 

So I think it's useful, but doesn't have to be necessarily visible by default.

**Comment by ebullient** [comment]
> Adding the properties feature is nice for the usecase of books/tutorials as you can then just share a command line and have a generated projected with everything but the code setup to go.
> 
> So I think it's useful, but doesn't have to be necessarily visible by default.

If it is there / shared / used in demos, it must be visible in help. 

Is there a good reason (other than convenience) to skip over setting configuration values in books/tutorials? Config is part of your code / part of your app .. especially with how much Quarkus puts there. Surely that is worth the time to explain and shouldn't be glossed over? What am I missing.

**Comment by maxandersen** [comment]
@agoncal describes his usecase at https://github.com/quarkusio/quarkus/issues/12924

**Comment by ebullient** [comment]
That isn't a use case. It is a "it would be nice... "

We can do a lot of things that would be nice, and when you do them all, you have a mess. :-P

**Comment by ebullient** [comment]
Updates after talking w/ Alexey.
Specifying the version in project `create` commands has been significantly changed. Also made `version` a subcommand on its own.

[cli-2021-05-12.txt](https://github.com/quarkusio/quarkus/files/6467510/cli-2021-05-12.txt)

**Comment by ebullient** [comment]
Additional changes: output-directory, specifying a package name, and enabling/disabling the build tool wrapper
[cli-2021-05-14.txt](https://github.com/quarkusio/quarkus/files/6480912/cli-2021-05-14.txt)

**Comment by ebullient** [comment]
Drafted a mixin for working with the registry-based extension catalog (with fallback)
```
      --[no-]registry-client
                            Use the Quarkus extension catalog
```

[cli-2021-05-17.txt](https://github.com/quarkusio/quarkus/files/6497664/cli-2021-05-17.txt)

File: devtools/cli/src/main/java/io/quarkus/cli/create/IncludeExampleCodeGroup.java
Comment (ia3andy) [inline devtools/cli/src/main/java/io/quarkus/cli/create/IncludeExampleCodeGroup.java:9]:
```suggestion
            "--no-code" }, description = "Include starter code provided by extensions or generate an empty project", negatable = true)
```

Comment (ia3andy) [reply devtools/cli/src/main/java/io/quarkus/cli/create/IncludeExampleCodeGroup.java:9]: So you understand, we stopped using the term "example code" for extensions provided code, the generated code is not meant to be deleted anymore but to get started from.

We can remove the `--example` option for now (this is meant for a different usecase but let's see if it's requested)

**Review by ia3andy — CHANGES_REQUESTED** [review]

**Review by aloubyansky — APPROVED** [review]
This looks good to me. Let's merge this after the Alpha3 has been tagged and @ia3andy's remarks have been addressed.

File: devtools/cli/src/main/java/io/quarkus/cli/create/IncludeExampleCodeGroup.java
Comment (ebullient) [reply devtools/cli/src/main/java/io/quarkus/cli/create/IncludeExampleCodeGroup.java:9]: ok. Was using old terminology from the CLI/Maven repo as I could see it

**Review by ia3andy — APPROVED** [review]
Cool :) & don't forget about the app config or I will add it back myself

**Comment by ebullient** [comment]
> Cool :) & don't forget about the app config or I will add it back myself

I told you I wanted to wait to add it back. I'd like to try a config command instead (need to get the other bits done first).

**Comment by maxandersen** [comment]
> I told you I wanted to wait to add it back. I'd like to try a config command instead (need to get the other bits done first).

the whole point of specifying config at create time is to not require additional commands but just have the project in the proper state at creation time leaving no or at least less room for error.

we can still consider editing commands in future but I see these as quite complementary usecases.

**Comment by maxandersen** [comment]
adding comments as I find things:

`qs` standalone just prints complaint and then tell user to run `qs --help` would be nicer if it just complained and then showed the help.

**Comment by maxandersen** [comment]
not sure if this is the behaviour we want/need but if you run `qs create` with `.env` file in the same dir you get output like:

```
qssnap create
2021-05-25 15:47:57,605 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.github.app.private.key" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
2021-05-25 15:47:57,606 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.github.app.webhook.secret" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
2021-05-25 15:47:57,606 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.github.app.webhook.proxy.url" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
2021-05-25 15:47:57,606 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.github.app.app.id" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
No subcommand specified, creating an app (see --help).
``` 

the `.env` file is:

```
QUARKUS_GITHUB_APP_APP_ID=
QUARKUS_GITHUB_APP_WEBHOOK_SECRET=
QUARKUS_GITHUB_APP_WEBHOOK_PROXY_URL=
QUARKUS_GITHUB_APP_PRIVATE_KEY=
```

I kinda get why its being picked up but would be nice somehow to know as a user why its complaining.

**Comment by maxandersen** [comment]
`qs create --help` does not show the help for `qs create run` ...should it since its the default ?

**Comment by maxandersen** [comment]
`qs --dryrun` seem to make no difference. project always created in a directory.

**Comment by maxandersen** [comment]
I think jbang template would be better of it created something called `src/main.java` then `jbang src` will just pick that no matter what additional files are in that directory as the "root"

**Comment by ebullient** [comment]
> `qs create --help` does not show the help for `qs create run` ...should it since its the default ?

I don't know what `qs create run` is.

**Comment by maxandersen** [comment]
```
$ qssnap create -a wonkabear --gradle
$ cd wonkabear
$ qssnap build
java.lang.NullPointerException
	at java.base/java.util.ArrayDeque.addAll(ArrayDeque.java:323)
	at io.quarkus.cli.build.GradleRunner.prepareBuild(GradleRunner.java:115)
	at io.quarkus.cli.Build.call(Build.java:33)
	at io.quarkus.cli.Build.call(Build.java:17)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2346)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2311)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at io.quarkus.cli.QuarkusCli.run(QuarkusCli.java:44)
	at io.quarkus.cli.QuarkusCli_ClientProxy.run(QuarkusCli_ClientProxy.zig:129)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:123)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:66)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:42)
	at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:30)
```

**Comment by maxandersen** [comment]
`qs build` does not work in a jbang project and its spending a surprising lot of time to in the end print:

`Was not able to find a build file in: /Users/max/code/personal/demo/wj/code-with-quarkus/src/wj/code-with-quarkus`

**Comment by maxandersen** [comment]
> I don't know what `qs create run` is.

`qs create app` sorry :)

**Comment by maxandersen** [comment]
```
$ qssnap ext list
quarkus-hibernate-orm-panache
quarkus-resteasy
```

but `qssnap rm hibernate` fails with:
```
qssnap ext rm hibernate
[ERROR] ❗ Multiple extensions matching 'hibernate'
     * io.quarkus:quarkus-hibernate-orm
     * io.quarkus:quarkus-hibernate-envers
     * io.quarkus:quarkus-hibernate-reactive
     * io.quarkus:quarkus-hibernate-orm-panache
     * io.quarkus:quarkus-hibernate-reactive-panache
     * io.quarkus:quarkus-hibernate-orm-panache-kotlin
     * io.quarkus:quarkus-hibernate-orm-rest-data-panache
     * io.quarkus:quarkus-hibernate-search-orm-elasticsearch
     * io.quarkus:quarkus-hibernate-validator
     try using the exact name or the full GAV (group id, artifact id, and version).
```

This is how old behavior was - still wrong though as in the list of project extensions rm hibernate should be trivial when there is just one such extension present.

**Comment by gsmet** [comment]
There are still tests failing, I won't merge it for CR1.

**Comment by ebullient** [comment]
y.. we'll sort the jbang ones today. Arguably, I could have left jbang support out..

File: devtools/cli/src/main/java/io/quarkus/cli/build/JBangRunner.java
Comment (maxandersen) [inline devtools/cli/src/main/java/io/quarkus/cli/build/JBangRunner.java:19]: its jbang.cmd or jbang.ps1 - no .bat's.

**Review by maxandersen — APPROVED** [review]
lgtm - the changes are local for cli and overall is better than what we have now.

**Comment by glefloch** [comment]
@ebullient I pushed a commit on your branch which fix the test. It just update the log level of the `GradleMessageWriter` to `lifecycle` which is the default level. I really don't know how this is currently working..

**Comment by ebullient** [comment]
> I really don't know how this is currently working..

I reached that point, but figured you and I could sort it out next. 😝

**Comment by Sgitario** [comment]
I tried to play with this PR and the changes look really promising!
I only wanted to spot one issue related to the `build` command, when using the build command and providing parameters (for example: `quarkus build -D=property=key`, it fails with:

```
[ERROR] Unknown lifecycle phase "property=key". You must specify a valid lifecycle phase
```

Note that I'm not sure if the parameter should be `-D=property=key` as stated in the help (perhaps a mispelling?) or it should be `-Dproperty=key`. However, I tried with both and it's also failing with the same error.
cc @ebullient

File: devtools/cli/src/main/java/io/quarkus/cli/build/MavenRunner.java
Comment (Sgitario) [inline devtools/cli/src/main/java/io/quarkus/cli/build/MavenRunner.java:118]: This is duplicated

**Comment by maxandersen** [comment]
i was just going to use alpha3 project create and noticed we only have this:

```
Quarkus version
  -p, --platform-bom=groupId:artifactId:version
                            A specific Quarkus platform BOM,
                            e.g. io.quarkus:quarkus-bom:1.13.4.Final
```

that is quite tedious compared to what we can do with maven/gradle plugins where we can specify the various parts on its own. i.e. i would expect to just do -pv 2.0.0.Alpha3 or -pg com.redhat.quarkus.platform instead of the full one.

**Comment by maxandersen** [comment]
alternatively we could allow -p ::2.0.0.Alpha3 ,i .e. partial GAV but not sure if thats too cryptic

**Comment by ebullient** [comment]
We are shooting for streams asap... the pom is then the less-used fallback

File: devtools/cli/src/main/java/io/quarkus/cli/build/MavenRunner.java
Comment (ebullient) [reply devtools/cli/src/main/java/io/quarkus/cli/build/MavenRunner.java:118]: I will add this to the list of follow on nit issues... i tested this at some point and it worked, but I realize I didn't add it to the actual test.. (the auto-generated help for maps does show that equals sign.. it should work either way)

**Comment by maxandersen** [comment]
Stream has exact same needs. Should not need to specify it all.

**Comment by ebullient** [comment]
> Stream has exact same needs. Should not need to specify it all.

my understanding (from Alexey) was that streams were more of 

> I tried to play with this PR and the changes look really promising!
> I only wanted to spot one issue related to the `build` command, when using the build command and providing parameters (for example: `quarkus build -D=property=key`, it fails with:
> 
> ```
> [ERROR] Unknown lifecycle phase "property=key". You must specify a valid lifecycle phase
> ```
> 
> Note that I'm not sure if the parameter should be `-D=property=key` as stated in the help (perhaps a mispelling?) or it should be `-Dproperty=key`. However, I tried with both and it's also failing with the same error.
> cc @ebullient
> 
> Update: the same happens with the `dev` command (quarkus dev -Dproperty=key)

I will add this to the list of follow on nit issues... i tested this at some point and it worked, but I realize I didn't add it to the actual test.. (the auto-generated help for maps does show that equals sign.. it should work either way)

**Comment by rsvoboda** [comment]
Hi, I miss any documentation related to the new CLI. Is there anything on the way?

**Comment by ebullient** [comment]
YES! in flight now. ;)

**Comment by ebullient** [comment]
> not sure if this is the behaviour we want/need but if you run `qs create` with `.env` file in the same dir you get output like:
> 
> ```
> qssnap create
> 2021-05-25 15:47:57,605 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.github.app.private.key" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
> 2021-05-25 15:47:57,606 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.github.app.webhook.secret" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
> 2021-05-25 15:47:57,606 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.github.app.webhook.proxy.url" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
> 2021-05-25 15:47:57,606 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.github.app.app.id" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
> No subcommand specified, creating an app (see --help).
> ```
> 
> the `.env` file is:
> 
> ```
> QUARKUS_GITHUB_APP_APP_ID=
> QUARKUS_GITHUB_APP_WEBHOOK_SECRET=
> QUARKUS_GITHUB_APP_WEBHOOK_PROXY_URL=
> QUARKUS_GITHUB_APP_PRIVATE_KEY=
> ```
> 
> I kinda get why its being picked up but would be nice somehow to know as a user why its complaining.

This is related to #19549

**Pull request merged**
